assets/js/modules/analytics-4/datastore/report.js

```diff
@@ -0,0 +1,175 @@
+import { MODULES_ANALYTICS_4 } from './constants';
+import { stringifyObject } from '../../../util';
+import {
+	isValidDateRange,
+	isValidOrders,
+} from '../../../util/report-validation';
+import {
+	normalizeReportOptions,
+	isValidDimensionFilters,
+	isValidDimensions,
+	isValidMetrics,
+} from '../utils';
+
+const fetchGetReportStore = createFetchStore( {
+	baseName: 'getReport',
+	controlCallback: ( { options } ) => {
+		return API.get(
+			'modules',
+			'analytics-4',
+			'report',
+			normalizeReportOptions( options )
+		);
+	},
+	reducerCallback: ( state, report, { options } ) => {
+		return {
+			...state,
+			reports: {
+				...state.reports,
+				[ stringifyObject( options ) ]: report,
+			},
+		};
+	},
+	argsToParams: ( options ) => {
+		return { options };
+	},
+	validateParams: ( { options } = {} ) => {
+		invariant(
+			isPlainObject( options ),
+			'Options for Analytics 4 report must be an object.'
+		);
+		invariant(
+			isValidDateRange( options ),
+			'Either date range or start/end dates must be provided for Analytics 4 report.'
+		);
+
+		const { metrics, dimensions, dimensionFilters, orderby } =
+			normalizeReportOptions( options );
+
+		invariant(
+			metrics.length,
+			'Requests must specify at least one metric for an Analytics 4 report.'
+		);
+		invariant(
+			isValidMetrics( metrics ),
+			'Metrics for an Analytics 4 report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "expression" and "alias" properties.'
+		);
+
+		if ( dimensions ) {
+			invariant(
+				isValidDimensions( dimensions ),
+				'Dimensions for an Analytics 4 report must be either a string, an array of strings, an object, an array of objects or a mix of strings and objects. If an object is used, it must have "name" property.'
+			);
+		}
+
+		if ( dimensionFilters ) {
+			invariant(
+				isValidDimensionFilters( dimensionFilters ),
+				'Dimension filters must be a map of dimension names as keys and dimension values as values.'
+			);
+		}
+
+		if ( orderby ) {
+			invariant(
+				isValidOrders( orderby ),
+				'Orders for an Analytics report must be either an object or an array of objects where each object should have "fieldName" and "sortOrder" properties.'
+			);
+		}
+	},
+} );
+
+const baseInitialState = {
+	reports: {},
+};
+
+const baseResolvers = {
+	*getReport( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const existingReport = registry
+			.select( MODULES_ANALYTICS_4 )
+			.getReport( options );
+
+		// If there is already a report loaded in state, consider it fulfilled
+		// and don't make an API request.
+		if ( existingReport ) {
+			return;
+		}
+
+		yield fetchGetReportStore.actions.fetchGetReport( options );
+	},
+};
+
+const baseSelectors = {
+	/**
+	 * Gets an Analytics report for the given options.
+	 *
+	 * @since 1.13.0
```

Comment:
There's a couple of these throughout, but these should all be using the appropriate placeholder since everything here is new.